src/phase_change/AllenCahnHierarchyIntegrator.cpp

```diff
@@ -0,0 +1,1824 @@
+    }
+
+    if (!(d_lf_convective_op_type == "CUI_CONSERVATIVE"))
+    {
+        TBOX_ERROR(d_object_name << "::AllenCahnHierarchyIntegrator():\n"
+                                 << " current implementation supports only\n"
+                                 << " CUI CONSERVATIVE limiter for Allen-Cahn equation\n");
+    }
+
+    if (input_db->keyExists("lf_solver_type"))
+    {
+        d_lf_solver_type = input_db->getString("lf_solver_type");
+        if (input_db->keyExists("lf_solver_db")) d_lf_solver_db = input_db->getDatabase("lf_solver_db");
+    }
+    if (!d_lf_solver_db) d_lf_solver_db = new MemoryDatabase("lf_solver_db");
+
+    if (input_db->keyExists("lf_precond_type"))
+    {
+        d_lf_precond_type = input_db->getString("lf_precond_type");
+        if (input_db->keyExists("lf_precond_db")) d_lf_precond_db = input_db->getDatabase("lf_precond_db");
+    }
+    if (!d_lf_precond_db) d_lf_precond_db = new MemoryDatabase("lf_precond_db");
+
+    return;
+} // AllenCahnHierarchyIntegrator
+
+void
+AllenCahnHierarchyIntegrator::initializeHierarchyIntegrator(Pointer<PatchHierarchy<NDIM> > hierarchy,
+                                                            Pointer<GriddingAlgorithm<NDIM> > gridding_alg)
+{
+    if (d_integrator_is_initialized) return;
+
+    // Perform hierarchy initialization operations common to all implementations
+    // of PhaseChangeHierarchyIntegrator.
+    PhaseChangeHierarchyIntegrator::initializeHierarchyIntegrator(hierarchy, gridding_alg);
+
+    // Operators and solvers are maintained for each variable registered with the
+    // integrator.
+    if (d_lf_solver_type == CCPoissonSolverManager::UNDEFINED)
+    {
+        d_lf_solver_type = CCPoissonSolverManager::DEFAULT_KRYLOV_SOLVER;
+    }
+    if (d_lf_precond_type == CCPoissonSolverManager::UNDEFINED)
+    {
+        const int max_levels = gridding_alg->getMaxLevels();
+        if (max_levels == 1)
+        {
+            d_lf_precond_type = CCPoissonSolverManager::DEFAULT_LEVEL_SOLVER;
+        }
+        else
+        {
+            d_lf_precond_type = CCPoissonSolverManager::DEFAULT_FAC_PRECONDITIONER;
+        }
+        d_lf_precond_db->putInteger("max_iterations", 1);
+    }
+
+    d_lf_solver = getAllenCahnEquationHelmholtzSolver(d_lf_var);
+    d_lf_rhs_op = getAllenCahnEquationHelmholtzRHSOperator(d_lf_var);
+
+    // Register additional variables required for present time stepping algorithm.
+    const IntVector<NDIM> cell_ghosts = CELLG;
+    const IntVector<NDIM> side_ghosts = SIDEG;
+    const IntVector<NDIM> face_ghosts = FACEG;
+    const IntVector<NDIM> no_ghosts = NOGHOSTS;
+
+    int lf_F_current_idx, lf_F_scratch_idx, lf_F_new_idx;
+    registerVariable(lf_F_current_idx,
+                     lf_F_new_idx,
+                     lf_F_scratch_idx,
+                     d_lf_F_var,
+                     cell_ghosts,
+                     "CONSERVATIVE_COARSEN",
+                     "CONSERVATIVE_LINEAR_REFINE");
```

Comment:
This seems wrong - these variables are registered but the indices will go out of scope and thus we can never use them. What's going on here?

Reply:
I have corrected this for all the variables. Please take a look.